plugins/plugins/sentinel/filter.go

```diff
@@ -0,0 +1,150 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sentinel
+
+import (
+	"fmt"
+	"strings"
+
+	sentinel "github.com/alibaba/sentinel-golang/api"
+	"github.com/alibaba/sentinel-golang/core/base"
+
+	"mosn.io/htnn/api/pkg/filtermanager/api"
+	types "mosn.io/htnn/types/plugins/sentinel"
+)
+
+func factory(c interface{}, callbacks api.FilterCallbackHandler) api.Filter {
+	return &filter{
+		callbacks: callbacks,
+		config:    c.(*config),
+		ctx:       make(map[string]interface{}),
+	}
+}
+
+type filter struct {
+	api.PassThroughFilter
+
+	callbacks api.FilterCallbackHandler
+	config    *config
+	ctx       map[string]interface{}
+}
+
+func (f *filter) DecodeHeaders(headers api.RequestHeaderMap, endStream bool) api.ResultAction {
+	res := f.getSource(f.config.GetResource(), headers)
+	if res == "" {
+		return api.Continue
+	}
+	api.LogDebugf("traffic control by: %s", res)
+
+	var attachments = make(map[interface{}]interface{})
+	for _, a := range f.config.attachments {
+		if a == nil || a.GetKey() == "" {
+			continue
+		}
+
+		v := f.getSource(a, headers)
+		if v == "" {
+			continue
+		}
+
+		attachments[a.GetKey()] = v
+	}
+
+	e, b := sentinel.Entry(res, f.config.params, sentinel.WithAttachments(attachments))
+
+	if b != nil {
+		api.LogDebugf("blocked, resource: %s, type: %s, rule: %+v, snapshot: %+v",
+			res, b.BlockType().String(), b.TriggeredRule(), b.TriggeredValue())
+
+		resp := &types.BlockResponse{
+			Message:    "sentinel traffic control",
+			StatusCode: 429,
+		}
+
+		switch b.BlockType() {
+		case base.BlockTypeFlow:
+			if r, exist := f.config.m.f[res]; exist && r.GetBlockResponse() != nil {
+				resp = r.GetBlockResponse()
+			}
+		case base.BlockTypeHotSpotParamFlow:
+			if r, exist := f.config.m.hs[res]; exist && r.GetBlockResponse() != nil {
+				resp = r.GetBlockResponse()
+			}
+		case base.BlockTypeCircuitBreaking:
+			if r, exist := f.config.m.cb[res]; exist && r.GetBlockResponse() != nil {
+				resp = r.GetBlockResponse()
+			}
+		}
+
+		header := make(map[string][]string)
+		for k, v := range resp.Headers {
+			vals := strings.Split(v, ",")
+			for i := range vals {
+				vals[i] = strings.TrimSpace(vals[i])
+			}
+			header[k] = vals
+		}
+
+		return &api.LocalResponse{
+			Code:   int(resp.StatusCode),
+			Msg:    resp.Message,
+			Header: header,
+		}
+	}
+
+	f.ctx["_entry"] = e
+	api.LogDebugf("passed, resource: %s", res)
+
+	return api.Continue
+}
+
+func (f *filter) EncodeHeaders(headers api.ResponseHeaderMap, endStream bool) api.ResultAction {
+	e, exist := f.ctx["_entry"].(*base.SentinelEntry)
+	if !exist {
+		return api.Continue
+	}
+	defer e.Exit()
```

Comment:
Thanks for pointing out the problem, but I don't know much about Websocket, can you expand on it or provide references.

Reply:
I guess the Exit will decrease the request count (please correct me if I am wrong). When handling streaming request like Websocket, the existence of a response header doesn't mean the end of a request. There will be a long time to process the response body after the header is sent.

Reply:
Yeah, you're right. So should it exit at EncodeData? Or is there a better time?

Reply:
What about putting it in the OnLog method?

Reply:
Thanks for your advice. I'll do it.

Reply:
After further consideration, I think traffic control only needs to pay attention to the ingress traffic.  The reason why I need to Exit in the response phase is that I need to obtain the response header for some processing, rather than wait for the completion of the response before Exit, so I think I should Exit in EncodeHeaders

Reply:
I see. Would you add some comments about the Exit?

Reply:
OK

Reply:
这里我认为还是应该放到 `Onlog` 里面去执行 `Exit()`.
sentinel 本意就是把响应发送给 downstream 后才去调用 `Exit()`.
 
在 client 断连的这种场景中，也可能并不能执行到  `EncodeData()` 阶段。
@WeixinX  
Onlog 方法还存在并发问题，请参考文档：site/content/zh-hans/docs/developer-guide/get_involved.md

Reply:
@alpha-baby 

我粗浅理解：关于指标的统计是在 `DecodeHeaders` 中获取 `api.Entry` 时执行的，也就是说在获取到 Entry 时就已经达到 Flow、HotSpot 规则的流控目的了，而 `SentinelEntry.Exit` 是 CircuitBreaker 规则需要通过 backend 响应码做指标统计。

我对上述意见有些疑问：
- 若无法执行 `EncodeData`，`Onlog` 还会被执行到吗？
- 既然已经断连了，CircuitBreaker 规则还需要统计状态码吗？

非常感谢你的答复！

Reply:
类似的插件 `limit count redis` 也没有关于 Onlog 的处理逻辑

Reply:
@WeixinX 
> 若无法执行 EncodeData，Onlog 还会被执行到吗？
Onlog 是每次都会执行的，
> 既然已经断连了，CircuitBreaker 规则还需要统计状态码吗？
那就不需要了

Reply:
@alpha-baby 所以实际上在 `DecodeHeaders` 做 `Exit` 操作也是没问题的，原因是：
- 正常情况下，DecodeHeaders 一定会被执行
- client 中断的情况下，DecodeHeaders 不一定会被执行，但此时已经不需要对 CircuitBreaker 规则做指标统计了，因为它依赖 backend 的响应码

Reply:
> 类似的插件 `limit count redis` 也没有关于 Onlog 的处理逻辑

我理解是否需要在 Onlog 中处理 sentinel.Exit 还是和流控策略先关的。你说的这个  `limit count redis` 插件里面本身就不需要处理类似 `Exit` 逻辑。我理解的有问题吗 @spacewander

Reply:
> 我粗浅理解：关于指标的统计是在 `DecodeHeaders` 中获取 `api.Entry` 时执行的，也就是说在获取到 Entry 时就已经达到 Flow、HotSpot 规则的流控目的了，而 `SentinelEntry.Exit` 是 CircuitBreaker 规则需要通过 backend 响应码做指标统计。

正常逻辑是没问题的，但是代码需要考虑各种边界情况。需要在 Onlog 中调用 `sentinel.Exit`  的最主要原因是因为在边界情况会出现 Exit 没有被调用，然后再 sentinel-golang 内部出现内存泄露

Reply:
> > 类似的插件 `limit count redis` 也没有关于 Onlog 的处理逻辑
> 
> 我理解是否需要在 Onlog 中处理 sentinel.Exit 还是和流控策略先关的。你说的这个 `limit count redis` 插件里面本身就不需要处理类似 `Exit` 逻辑。我理解的有问题吗 @spacewander

limit count redis 算的是给定时间段里的请求数，不是并发，不需要在结束请求时减去值

Reply:
> Onlog 方法还存在并发问题

@alpha-baby，我不太理解这个并发问题什么意思？请麻烦解释下 ~